Add link UI with autocomplete to the image block

jorgefilipecosta opened the pull request:
## Description
Fix: https://github.com/WordPress/gutenberg/issues/8265

This PR implements a button in the image block that allows users to use the link UI already available in the paragraph.
Some components were extracted from the link format, to make them available in the image block.

## How has this been tested?
I added an image block.
I selected an image.
I pressed the link button in the toolbar.
I tried to add a link, I verified I can use autocomplete; I selected a link; I verified the link was applied.
I verified in the block inspector that the "Link To" field was changed to "Custom Url".

jorgefilipecosta committed:
Add link UI with auto complete to image block.

ajitbohra reviewed (commented):
Functionality wise works as advertised :+1: 

One observation when switching option from media/attachment to custom it uses the value of media/attachment.

youknowriad commented:
I think we went back and forth on this one in previous releases.
It works well now but I wonder about the UI. It feels inconsistent to have some links settings in the inspector and others in the sidebar. I pinged some design eyes.

jasmussen commented:
Thanks for this, very nice work. Overall GIF:

![linkui](https://user-images.githubusercontent.com/1204802/57611165-86373f80-7572-11e9-811e-8f6c2a9fa9f5.gif)

The concern that Riad highlights is, correct me if I'm wrong, summarized in this GIF:

![sidebar](https://user-images.githubusercontent.com/1204802/57611193-90f1d480-7572-11e9-870e-a9f91206c0d0.gif)

The concern is that it's duplicate UI. Usually when there are two ways to accomplish the same, it begs a question of the user: which way is the right way?

Note that this is different from _redundant UI_, which is also duplicate, but is identical.

This challenge is exactly why I was a bit sad to see the old image link UI being moved to the sidebar as part of what is currently in master. I knew that it was a temporary stopgap to ensure feature parity, but that the UI implemented was not very good because it had to happen fast. This is where it bites us. Not saying that to disparage, just as context — I've been around for the ride, and sometimes you have to take detours, and I'm happy we're now looking at it again.

To resurrect that discussion a little bit — linking an image feels like a basic operation. Following our principles, that makes it primary UI, which means it should not be in the sidebar. For that reason, I quite like that it's now a primary operation, and the fact that it's similar to the paragraph UI is good. Technically the two also work nicely side by side — one is aware of the other and you can go about your day. As such, I have no objection to this PR being merged as is, so long as we ticket improvements to the sidebar link UI. 

However those sidebar improvements are important. As a strawman, this seems like it would be an improvement:

![mockup](https://user-images.githubusercontent.com/1204802/57611730-c64af200-7573-11e9-8ffe-5ac4e6aa170f.png)

At least then, the two UI's would be unified into one. 

---

I also noticed something interesting: 

<img width="333" alt="Screenshot 2019-05-13 at 11 28 27" src="https://user-images.githubusercontent.com/1204802/57611772-db278580-7573-11e9-9ef2-3e3bcf530a31.png">

Notice the extra small 1px gray border around the formatting buttons — that's a regression. It's also in master, so it's not the fault of this branch, but it is just more visible here because the button pops out a popover and the extra border lingers there. I'll investigate.

jasmussen commented:
PR for the lingering outline thing in https://github.com/WordPress/gutenberg/pull/15592

talldan commented:
> However those sidebar improvements are important. As a strawman, this seems like it would be an improvement

Some thoughts on the design ... The user's interaction with with the link popover naturally flows:
1. Enter a URL
2. Change the settings for that URL (Opens in new tab etc.)

With the 'Link To' option that order doesn't make sense because 'Link To' modifies the URL that was initially entered. In the sidebar's link settings, it's the first step, which makes sense to me. I wonder if there's another way to express that setting in the popover version. 🤔

jasmussen commented:
> With the 'Link To' option that order doesn't make sense because 'Link To' modifies the URL that was initially entered. In the sidebar's link settings, it's the first step, which makes sense to me. I wonder if there's another way to express that setting in the popover version that would allow the user to have a more natural editing flow. 🤔

This is a good point, and speaks to the existing limitations of the legacy UI. It definitely needs a re-think overall. I don't think additional buttons are the solution — in my mind "attachment page" and "media file" have more in common with auto-complete suggestions than they do modal options for the entire link interface. They're all just links. My suggestion for putting it in the advanced dropdown for now was more to at least group the link UIs together so we can keep moving. 

@kjellr do you have any magical insights here that eldude me? There could be some simple elegant solution I'm not seeing.

jorgefilipecosta commented:
What if we add the Link to option to the popover before the URL input? The open in a new tab would still be an andavanced option.

jasmussen commented:
Noting that if you rebase, that lingering focus thing is gone :) — if not, it will be gone post merge. 

> What if we add the Link to option to the popover before the URL input? The open in a new tab would still be an andavanced option.

Can you elaborate a little bit?

As mentioned, I don't feel it necessary to re-do the link UI to be able to ship THIS PR, if we mean to ship it quickly. Just ticket the improvement and we can move. 

However if there _is_ bandwidth for a re-think, I would suggest that Google Docs has a solid interface:

![docs](https://user-images.githubusercontent.com/1204802/57619800-239c6e80-7587-11e9-9f28-5309b1cb7743.gif)

It is a single link UI, but it also shows contextual options at the bottom of the dropdown. Just press tab in that link dialog to access those options.

The equivalent for us is almost there already: you can search for an existing page to link to, and an "autocomplete-esque" UI will pop out. We could do so that when you invoke the link dialog on the image, that autocomplete UI is already open, and shows two options at the bottom: link to attachment page, or link to media file. The keyboard flow would be this:

- Tab to link button on toolbar, press enter. 
- Focus is set in inputfield. Autocomplete box with the additional options are already open.
- Type URL and autocomplete box closes. Press enter to apply URL.
- OR: type a page or post title to invoke the search. Use arrow keys to choose page, press enter to apply link to page or post. 
- OR: use arrow keys or press tab right off the bat to choose "link to media file" or "link to attachment page". 

Does that make sense? Need a mockup?

kjellr commented:
> The equivalent for us is almost there already: you can search for an existing page to link to, and an "autocomplete-esque" UI will pop out. We could do so that when you invoke the link dialog on the image, that autocomplete UI is already open, and shows two options at the bottom: link to attachment page, or link to media file. 

Yep, I fully endorse this. 🙌 I actually mocked this up a while back and just dug it up: 

![Frame](https://user-images.githubusercontent.com/1202812/57620520-00f26180-7557-11e9-9197-92d545028668.png)

Also, is it possible for us to be smart and say "Link to _image_ file" here instead of "media file"? I think that will be more clear to folks.

jasmussen commented:
Killer mockup!

I do think a separator and some icons next to those two options will give them slightly more presence, but the idea feels solid.

kjellr commented:
Yeah — I figured we'd still be using the same component we're used to here, right? If so, we don't usually put icons in there.

For the separator, were you thinking that'd be a separator between these and other options in the dropdown (once they appear)? I'd figured that these options would just disappear once you typed something else, since they're not technically auto-completes.

jasmussen commented:
An absent icon wouldn't block anything, but it would make it less "missable" — right now it's a little invisible down there. 

> For the separator, were you thinking that'd be a separator between these and other options in the dropdown (once they appear)? I'd figured that these options would just disappear once you typed something else, since they're not technically auto-completes.

Good point. Yes those options available right off the bat would disappear as soon as you type, so no need to separate them.

kjellr commented:
Here's a revised mockup with icons. Not totally sure what to use for the `Media file` one — theoretically it makes sense to use the image icon, but then it's also repeating the block icon nearby, which may be confusing. 

<img width="624" alt="Frame" src="https://user-images.githubusercontent.com/1202812/57640947-b508e200-7581-11e9-9fb2-980373646858.png">

jorgefilipecosta commented:
The UI was updated yesterday to follow the mockups and can be tested. I felt the need to add some UI elements like a button that allows the user to remove the link (now we don't have the option to set the link to none).
The duplicate UI was removed.
The only thing missing are the icons, yesterday I did not see them. I will try to add the icons soon.

kjellr commented:
This is looking pretty good so far. I'm seeing the suggested results appear up at the top of the screen when I click the toolbar icon though, which is a little weird: 

![link](https://user-images.githubusercontent.com/1202812/57695903-77f22d80-761d-11e9-97d2-d699fce4f28a.gif)

I wonder if it has to do with positioning? In any case, once you have the icons in there, I'm happy to jump in and do some super minor styling cleanup if you'd like — I noticed the hover state for the link icon doesn't match other toolbar icons for instance. 👍 

For the icons, the media file one is the same as the Image block icon. For the "page" icon, let's use [the `pages` icon from Gridicons](https://github.com/Automattic/gridicons/blob/master/svg-min/gridicons-pages.svg) it looks like other Gutenberg page icons are based on that one.

mapk commented:
Just tested this! I Love the combined experience. 

I saw the same positioning issue as Kjell [above](https://github.com/WordPress/gutenberg/pull/15570#issuecomment-492210243). 

I'd also suggest that we add more space in between the items in the dropdown too. Once we add icons, that might help. Similar spacing as this...

![Screen Shot 2019-05-15 at 7 02 17 AM](https://user-images.githubusercontent.com/617986/57781729-9c164300-76df-11e9-8a91-84e80786ee9d.png)

paaljoachim commented:
Just adding in a thought. As each image has linking in its own toolbar. At a later time this could be transferred over to a gallery to link each image separately. (But then again a gallery could be just nested single images placed together..:)

jorgefilipecosta commented:
Hi @kjellr, @mapk, @paaljoachim this PR suffered a big refactor to address the problems raised.
The problem of the suggestions appearing in a different position at the start happened because we had a PopOver inside another PopOver I tried to solve for the general case of having a popover inside another one but the solution would evolve adding much more complexity to the popover which is already complex. So I took a step back.
The PR now uses its own menu for these actions instead of using suggestions, it avoids the usage of a popover inside a popover, and it also solves the design problems raised about not looking like the other menus. I think the PR is ready for a new look.

mapk commented:
> The PR now uses its own menu for these actions instead of using suggestions, it avoids the usage of a popover inside a popover, and it also solves the design problems raised about not looking like the other menus. I think the PR is ready for a new look.

Thanks for working through this, @jorgefilipecosta! I'm currently testing this now. Here's my first observation that needs some styling love.

The toolbar link item doesn't have the same hover state as the other items.

![hover](https://user-images.githubusercontent.com/617986/58520235-b1717f80-816b-11e9-8aa0-d2c3920ff8c7.gif)

mapk commented:
There also seems to be a weird layering effect with the multiple dropdowns that appear. Check this out.

![drops](https://user-images.githubusercontent.com/617986/58520392-67d56480-816c-11e9-9a7a-72927a67d895.gif)


I really think when one dropdown is showing all others should disappear.

kjellr commented on packages/block-library/src/image/edit.js:
```diff
@@ -96,6 +108,154 @@ const isTemporaryImage = ( id, url ) => ! id && isBlobURL( url );
+	return (
+		<>
+			<IconButton
+				icon="admin-links"
```
Regarding @mapk's comment: 

> The toolbar link item doesn't have the same hover state as the other items.

This just needs a `components-toolbar__control` class to set that straight. 👍 

```suggestion
				icon="admin-links"
				className="components-toolbar__control"
```

kjellr commented:
> There also seems to be a weird layering effect with the multiple dropdowns that appear. Check this out.

Yeah, when the additional details are open, this happens on all the link popovers. I don't think that's a blocker for this PR, and we can sort it out globally as part #15872. 

We can and should prevent that in the default state here though. If we remove the bottom layer, it actually looks a little weird to my eyes: 

![Screen Shot 2019-05-29 at 1 33 44 PM](https://user-images.githubusercontent.com/1202812/58578850-c277ca80-8217-11e9-875a-faaf341090d8.png)

So I'd suggest we just make those default suggestions take up the full-width: 

![Screen Shot 2019-05-29 at 1 34 09 PM](https://user-images.githubusercontent.com/1202812/58578885-da4f4e80-8217-11e9-9ba6-a62bdffb2633.png)

That looks pretty natural to me. I'd also suggest we change the interior padding of the two suggested options so that they're only `$grid-size` (`8px`) on the left side. That way they'll line up nicely with the beginning of the text above them: 

![Screen Shot 2019-05-29 at 1 35 24 PM](https://user-images.githubusercontent.com/1202812/58578956-04a10c00-8218-11e9-82f5-75b1830311cc.png)

jorgefilipecosta commented:
Thank you for the reviews @mapk, and @kjellr the feedback was applied. I guess we should address https://github.com/WordPress/gutenberg/issues/15872 in other PR to avoid increasing the size of this one.

mapk commented:
> I guess we should address #15872 in other PR to avoid increasing the size of this one.

I agree. 

I just tested again and it's looking really good! One thing I noticed is that the padding under the last form field in the dropdown is really big. It should match the same padding around the rest of the dropdown.

![Screen Shot 2019-05-29 at 4 01 27 PM](https://user-images.githubusercontent.com/617986/58597157-95cdae00-822b-11e9-9445-e513c34fbe54.png)


I think if we add a `:last-child` psuedo-class on the `.components-base-control` to remove the `bottom-margin: 16px` then we should be good to go.

kjellr commented on packages/block-editor/src/components/url-popover/style.scss:
```diff
@@ -50,10 +62,31 @@
 	padding: $panel-padding;
 	border-top: $border-width solid $light-gray-500;
 
 	.components-base-control:last-child .components-base-control__field {
```
```suggestion
	.components-base-control:last-child,
	.components-base-control:last-child .components-base-control__field {
```

I think this should take care of @mapk's [comment below](https://github.com/WordPress/gutenberg/pull/15570#issuecomment-497140835).

Once this is fixed, I think this PR is good to go. 👍

jorgefilipecosta commented:
Thank you for the reviews @mapk, and @kjellr. I verified the suggestion provided by @kjellr works as expected 👍

kjellr commented:
Just gave this one last test, and I have a quick design question: when there is a link added, should the block use the "unlink" icon in the toolbar, like we do in the paragraph block?

<img width="721" alt="Screen Shot 2019-06-05 at 8 13 38 AM" src="https://user-images.githubusercontent.com/1202812/58955418-ef713380-8769-11e9-8924-570d1838be00.png">

Currently, the icon does not change: 

<img width="460" alt="Screen Shot 2019-06-05 at 8 13 12 AM" src="https://user-images.githubusercontent.com/1202812/58955437-fa2bc880-8769-11e9-9fe5-aea51bb16ec3.png">

jorgefilipecosta commented:
> Just gave this one last test, and I have a quick design question: when there is a link added, should the block use the "unlink" icon in the toolbar, like we do in the paragraph block?

Hi @kjellr, I don't we should follow the same approach, in the RichText formats, we can do that because when the cursor is on a link the popover is automatically open.  In the image case, I don't think the popover should automatically open when the image is selected. So if we added the unlink button if I added a link and then I wanted to change the link rel, I would need to click the unlink button the link would be removed, I would then click the link button add the link again and select the correct rel. That seems an un-optimal, that was the reason why I did not follow the unlink button approach and opted to add an explicit button to remove the link in the popover which is not available in the paragraph.

kjellr commented:
> In the image case, I don't think the popover should automatically open when the image is selected. So if we added the unlink button if I added a link and then I wanted to change the link rel, I would need to click the unlink button the link would be removed, I would then click the link button add the link again and select the correct rel. That seems an un-optimal, that was the reason why I did not follow the unlink button approach and opted to add an explicit button to remove the link in the popover which is not available in the paragraph.

That's great reasoning, thanks. 

One small a11y note: I expected to be able to be able to use the down arrow to get into the Media File and Attachment Page options. But tab worked fine instead, so I _think_ we're ok. I'd love a11y team feedback on that, and the updated UI in general.

afercia commented:
Discussed during today's accessibility bug scrub. At a first glance, anything that moves us away from the concept of “Sidebar” sounds good 🙂 

Needs to be tested for implementation details. Keeping the "Needs a11y feedback" label for now.

afercia commented:
Additional note: One more concern is consistency. Not just for accessibility, also for general usability. These “Edit” buttons do very different things across the various blocks: Image, Video, Audio, etc.

Not sure forcing users to learn a new UI each time is the best user experience. On the other hand, if this pattern turns out to work well, it could be considered also for other blocks.

afercia commented:
Discussed during today's accessibility bug scrub. At a first glance and looking at the GIFs this seems a step in the right direction, as it moves the UI link from the sidebar (not so accessible) to the block. Having the relevant UI pieces "in place" within the block sounds sensible.

Would need some testing with assistive technologies and some real user testing. WCEU could be a good opportunity.

jorgefilipecosta commented:
I did some tests on this PR with voice over and things worked as expected. I guess this PR should be ready merge.

kjellr reviewed (approved):
👍 This works well! Thanks, @jorgefilipecosta!

youknowriad commented on packages/block-editor/src/components/url-popover/index.js:
```diff
@@ -47,28 +60,107 @@ class URLPopover extends Component {
 	}
 }
 
+const LinkEditor = ( {
```
I think we need to reorganize all the link related comments better. For example why is this in the popover? It can be used without it right? Also the relationships between all these components is not obvious. we should probably have a README for each one of these components and organize them in a common folder or something like that.

Makes me wonder if we should mark all these things as "Experimental" before doing the reorganization.

jorgefilipecosta replied:
Yes, I think it may be possible to use these components outside the URLPopover, but they are precise UI pieces, and the most probable use case is inside a URLPopover component.
I don't think we should expose these components directly as part of wp.components otherwise we populate the module with particular UI pieces, I agree we should mark these components as experimental until we decide on a more profound reorganization.

aduth commented on packages/format-library/src/link/inline.js:
```diff
@@ -38,45 +31,6 @@ function isShowingInput( props, state ) {
 }
 
-const LinkEditor = ( { value, onChangeInputValue, onKeyDown, submitLink, autocompleteRef } ) => (
-	// Disable reason: KeyPress must be suppressed so the block doesn't hide the toolbar
```
While it's technically fair to say the comment was no longer necessary because the changes remove the need to disable the ESLint rule, there was some value in having an explanation here previously for _why_ we stop propagation. Considering the purpose of a specific `Event#stopPropagation` is almost never easily understandable at a glance, we should strive to retain as much information for their existence as possible.

Coincidentally, I see it as similar to why we have these "Disable reason:" comment, in that it's similarly non-obvious to a future maintainer why we disable an ESLint rule without a comment explaining the rationale (and, to an arguably greater effect, forcing ourselves to explain it might actually deter us from going through with hastily introducing an escape hatch).

For context: I arrived here after trying to determine whether I could remove the `stopKeyPropagation` in a refactor, because I can't understand based on the current code why it needs to exist.
